### PR TITLE
Implement training visualization utilities

### LINKED ISSUE #32: Implement training visualization utilities
Tools for visualization of following:
- performance
- policy
- action value
- episode
- model

### PR DESCRIPTION
**Pull request by JanKapala**
Closes #32

### CODE REVIEW

File: .ipynb_checkpoints/Untitled-checkpoint.ipynb [hunk 1]
```diff
@@ -0,0 +1,108 @@
+{
```
Comment (wujuu) [inline .ipynb_checkpoints/Untitled-checkpoint.ipynb]: this file is necessary?

File: .gitignore [hunk 2]
```diff
@@ -2,4 +2,4 @@
 __pycache__/
 training_platform/examples/agent0.ai
 training_platform/examples/agent1.ai
-
+game_app/settings.cfg
```
Comment (wujuu) [inline .gitignore]: nice @Klooskie what does this file do?

File: all_tests/environments/test_board.py [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
+
+REL_PROJECT_ROOT_PATH = "./../../"
+ABS_FILE_DIR = os.path.dirname(os.path.abspath(__file__))
+ABS_PROJECT_ROOT_PATH = os.path.normpath(os.path.join(ABS_FILE_DIR, REL_PROJECT_ROOT_PATH))
+sys.path.append(ABS_PROJECT_ROOT_PATH)
+# -------------------------PROJECT-ROOT-PATH-APPENDING----------------------END#
+
 from unittest import TestCase
 from environments.tic_tac_toe.tic_tac_toe_engine import _Board
+# from all_tests.mock.test_mock_action import MockAction
```
Comment (wujuu) [inline all_tests/environments/test_board.py]: commented import?

File: reinforcement_learning/__init__.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+from reinforcement_learning.base.base_agent import BaseAgent
+
+# Concrete agents
+from reinforcement_learning.agents.basic_mc_agent.basic_mc_agent import BasicAgent
+from reinforcement_learning.agents.random_agent.random_agent import RandomAgent
+from reinforcement_learning.agents.n_step_agent.n_step_agent import NStepAgent
+from reinforcement_learning.agents.q_learning_agent.q_learning_agent import QLearningAgent
+
+# Training class
+from reinforcement_learning.common.simple_training import SimpleTraining
```
Comment (wujuu) [inline reinforcement_learning/__init__.py]: are __init__ files really necessary?

File: reinforcement_learning/agents/basic_mc_agent/Digraph.gv [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	"1585b531-60f5-499a-a385-49cea5beb9e3" [label="#O
+OO"]
+	"7467df56-3140-439d-9898-8dc9816d088f" -> "1585b531-60f5-499a-a385-49cea5beb9e3"
+	node [shape=circle]
+	"dbd4246e-2d05-432e-b9e7-fef140ade039" [label="[0 0]"]
+	"1585b531-60f5-499a-a385-49cea5beb9e3" -> "dbd4246e-2d05-432e-b9e7-fef140ade039"
+	node [shape=diamond]
+	"6640b861-77ff-46a1-9f48-fac9a31ba331" [label=10]
+	"dbd4246e-2d05-432e-b9e7-fef140ade039" -> "6640b861-77ff-46a1-9f48-fac9a31ba331"
+}
```
Comment (wujuu) [inline reinforcement_learning/agents/basic_mc_agent/Digraph.gv]: maybe we shouldnt add this file to git

Comment (wujuu) [reply reinforcement_learning/agents/basic_mc_agent/Digraph.gv]: same below

File: reinforcement_learning/agents/basic_mc_agent/episode.py [hunk 6]
```diff
@@ -1,104 +1,116 @@
...
-#     e.append(r2)
-#
-#     s4 = MockState([[-1, 0], [0, 0]])
-#     a4 = MockAction([0, 0])
-#     e.append((s4, a4))
-#     r4 = 10
-#     e.append(r4)
-#
-#     print(e)
-#     e.view()
```
Comment (wujuu) [inline reinforcement_learning/agents/basic_mc_agent/episode.py]: finally deleted idiotic simple state

File: reinforcement_learning/agents/basic_mc_agent/episode.py [hunk 7]
```diff
@@ -1,104 +1,116 @@
...
+        return graph
+
+    def _repr_svg_(self):
+        return self._get_graph()._repr_svg_()
+
+    def view(self):
+        return self._get_graph().view()
+
+
+if __name__ == '__main__':
```
Comment (wujuu) [inline reinforcement_learning/agents/basic_mc_agent/episode.py]: we should delete this or add to tests

File: reinforcement_learning/agents/basic_mc_agent/returns.py [hunk 8]
```diff
@@ -84,43 +83,42 @@ def view(self):
...
-#     r[s4, a4].append(G)
-#     G = -9.4
-#     r[s4, a4].append(G)
-#     G = -10
-#     r[s4, a4].append(G)
-#
-#     print(r)
-#     # r.view()
\ No newline at end of file
+if __name__ == '__main__':
```
Comment (wujuu) [inline reinforcement_learning/agents/basic_mc_agent/returns.py]: as above

File: reinforcement_learning/agents/basic_mc_agent/stochastic_model.py [hunk 9]
```diff
@@ -78,44 +77,44 @@ def _repr_svg_(self):
...
-#     m[s2, a4] = s6
-#     m[s2, a4] = s6
-#     m[s2, a4] = s6
-#     m[s2, a4] = s7
-#     m[s2, a4] = s7
-#     m[s2, a4] = s7
-#
-#     m.view()
\ No newline at end of file
+if __name__ == '__main__':
```
Comment (wujuu) [inline reinforcement_learning/agents/basic_mc_agent/stochastic_model.py]: as above

File: reinforcement_learning/agents/common/action_value_derived_policy.py [hunk 10]
```diff
@@ -45,31 +51,62 @@ def epsilon_greedy(self, state, action_space, epsilon=0.1):
...
-#
-#     egp = ActionValueDerivedPolicy(av)
-#
-#     print(egp)
-#     egp.view()
\ No newline at end of file
+        return self._get_graph().view()
+
+
+if __name__ == '__main__':
```
Comment (wujuu) [inline reinforcement_learning/agents/common/action_value_derived_policy.py]: as above

File: reinforcement_learning/agents/common/lazy_tabular_action_value.py [hunk 11]
```diff
@@ -93,19 +101,22 @@ def _repr_svg_(self):
...
-#     a3 = MockAction([1, 0])
-#
-#     av[s, a1] = 6
-#     av[s, a2] = 0.8
-#     av[s, a3] = -10
-#
-#     print(av)
-#     av.view()
\ No newline at end of file
+if __name__ == '__main__':
```
Comment (wujuu) [inline reinforcement_learning/agents/common/lazy_tabular_action_value.py]: as above1

File: reinforcement_learning/__init__.py (hunk 4 above)
Comment (JanKapala) [reply reinforcement_learning/__init__.py]: Yep

**Pull request merged**